Port to plotly.py v4_subplots and renderers

jonmmease opened the pull request:
## Overview
This PR updates plotly_express to use the plotly.py `plotly.subplots.make_subplots` functionality to build faceted figures, rather than the current `layout.grid` approach.  By building on `plotly.subplots.make_subplots`, it will be much easier for users to modify faceted figures in plotly.py.

It also removes `ExpressFigure` and relies on the plotly.py `plotly.io.renderers` configuration to display figures.

Because this PR requires plotly.py to be imported with the `v4_subplots` and `renderer_defaults` future flags.  If plotly_express is imported first then this is done automatically, but if plotly.py is import first, without these future flags, then an error will be raised when plotly_express is imported.

## Examples
Update a y-axis based on row
```python
import plotly_express as px
iris = px.data.iris()
fig = (px.scatter(iris, x="sepal_width", y="sepal_length", facet_row='species', range_x=[0, 5], log_y=True))
fig.update_yaxes(title_font_color='red', row=3)
```
![newplot-38](https://user-images.githubusercontent.com/15064365/58055213-09d0cd80-7b2b-11e9-82ce-6c072ae85d0c.png)

Update a trace based on row
```python
fig = (px.scatter(iris, x="sepal_width", y="sepal_length", facet_row='species', range_x=[0, 5], log_y=True))
fig.update_traces(marker_color='red', row=2)
```
![newplot-39](https://user-images.githubusercontent.com/15064365/58055269-35ec4e80-7b2b-11e9-8f20-2dd3dd543b52.png)

### Marginals and Faceting
Support for combining marginals and faceting
```python
px.scatter(iris, x="sepal_width", y="sepal_length", facet_row='species', marginal_y='histogram')
```
![newplot-40](https://user-images.githubusercontent.com/15064365/58055681-90d27580-7b2c-11e9-8356-2c9241e86e90.png)

```python
fig = px.scatter(tips, x="total_bill", y="tip", facet_row="time", facet_col="day", # color="smoker",
                 category_orders={"day": ["Thur", "Fri", "Sat", "Sun"], "time": ["Lunch", "Dinner"]},
                 marginal_y='rug', marginal_x='histogram', color='sex'
                )

fig.update(layout_height=800)
```
![newplot-43](https://user-images.githubusercontent.com/15064365/58055908-4ef5ff00-7b2d-11e9-9943-3d29928e540b.png)

## Faceting non-cartesian trace types
This PR doesn't add `facet_row`/`facet_col` to any additional plot types, but when these args are added, faceting will "work" automatically.  I didn't add it in this PR because the default spacing for other trace types can look pretty bad, so we'll need to work through how these other subplot types shoud look with faceting.

cc @nicolaskruchten

jonmmease commented on plotly_express/_core.py:
```diff
@@ -295,126 +306,139 @@ def configure_axes(args, constructor, fig, axes, orders):
     }
     if constructor in configurators:
-        fig.update(layout=configurators[constructor](args, fig, axes, orders))
+        configurators[constructor](args, fig, orders)
```
configurators now update the layout in place rather than returning a new layout

nicolaskruchten replied:
I don't mind (this is how it originally was!) but why this change now?

jonmmease replied:
This was because I wanted to be able to use `update_(x|y)axes` inside the `configure_cartesian_axes`.

nicolaskruchten replied:
makes sense!

jonmmease commented on plotly_express/_core.py:
```diff
@@ -295,126 +306,139 @@ def configure_axes(args, constructor, fig, axes, orders):
 
 
-def set_cartesian_axis_opts(args, layout, letter, axis, orders):
+def set_cartesian_axis_opts(args, axis, letter, orders):
```
`axis` is now a `graph_objs` instance rather than a string

nicolaskruchten commented on plotly_express/_core.py:
```diff
@@ -1,9 +1,15 @@
+from _plotly_future_ import renderer_defaults, v4_subplots
```
can we think of a not-too-nasty way to have this work just like Plotly.py wherein if a certain future flag is set, it uses this mode?

nicolaskruchten replied:
And/or just to make it easier to maintain, are there some structural changes here which aren't specific to v4 features that we should backport to `master` sooner to make it easier to maintain this branch?

jonmmease replied:
> can we think of a not-too-nasty way to have this work just like Plotly.py wherein if a certain future flag is set, it uses this mode?

This was my original goal, the changes/refactoring ended up being pretty dispersed through the nested hierarchy of functions so I didn't see an easy way to do it.  I'll take another look this evening and see if If can isolate the changes.

nicolaskruchten replied:
I'm still on the fence about whether pre-v4-release versions of PX should turn on the future flags. I kinda think they should :) and then we don't need to worry about it too much.

thoughts?

nicolaskruchten replied:
I think the "a bunch of pre-release testing from early adopters" is kind of cool

jonmmease replied:
Yeah, if it were easy I'd have plotly_express follow the future_flags.  But since I think it would be pretty intrusive, and more complex to maintain, I'm inclined to turn it on by default.

We could also consider keeping plotly_express in an alpha state until the release of plotly.py v4. This way the opt-in to the new behavior would be to install the alpha version.

nicolaskruchten commented on plotly_express/_core.py:
```diff
@@ -295,126 +287,141 @@ def configure_axes(args, constructor, fig, axes, orders):
+                showgrid=args["marginal_x"] == 'histogram',
+                row=row
+            )
+            fig.update_xaxes(
```
has this been run through `black` ? :)

jonmmease replied:
Done now

nicolaskruchten replied:
Thanks! I know the plotly.py repo doesn't have that in place but this one does for now ;)

jonmmease replied:
Yeah, I'm interested in using black in plotly.py too.  Have you looked into the git integration https://black.readthedocs.io/en/stable/version_control_integration.html?

nicolaskruchten commented on plotly_express/_core.py:
```diff
@@ -869,6 +913,42 @@ def make_figure(args, constructor, trace_patch={}, layout_patch={}):
+
+                # Find row for trace, handling facet_row and marginal_x
+                if m.facet == 'row':
+                    if has_marginal_x:
```
Is this case for marginals + facets? that's not supported right now in the grouping logic.

jonmmease replied:
Yes. With these changes I *think* marginals + facets are working. See the screenshot in the description above.

nicolaskruchten replied:
aha I see, that's interesting. Cool that it works! 

I think that for combining them, I was envisioning per-row and per-column marginals rather than per-subplot marginals, which isn't quite as straightforward (although not impossible, just requires a second group-by pass). Thoughts?

jonmmease replied:
I liked per subplot marginals, but I suppose per row/col could useful to. Would it make sense for this to be the default and then later add an option to change the behavior?

nicolaskruchten commented:
> I didn't add it in this PR because the default spacing for other trace types can look pretty bad, so we'll need to work through how these other subplot types shoud look with faceting.

can I see an example? Is the spacing too big or too small?

jonmmease commented:
Sure, here are some examples:
![newplot-45](https://user-images.githubusercontent.com/15064365/58173000-232b6400-7c68-11e9-9883-2b40d90ab12b.png)

![newplot-46](https://user-images.githubusercontent.com/15064365/58173036-3c341500-7c68-11e9-87d6-1c2236a1ec2e.png)

![newplot-47](https://user-images.githubusercontent.com/15064365/58173050-4e15b800-7c68-11e9-93e9-1a15dd99b90a.png)

I see two issues:
  - The first is that different subplot types will need different default padding.  This should just be a matter of taking the time to work through all of the subplot types.
  - The second is that subplot types with a fixed aspect ratio (ternary and polar) are centered in their domain, so the row/column labels don't line up well.  I'm not really sure what to do about this.

jonmmease commented:
@nicolaskruchten, I just updated the PR with master and removed the marginals per facet.

If faceting and marginal specified then faceting wins.  It's still
possible to facet columns and have marginals per columns or to facet rows
and have marginals per row.

e.g.
![newplot (8)](https://user-images.githubusercontent.com/15064365/59453237-7911a880-8ddd-11e9-904e-3e48d6c7ecb3.png)

If this looks alright, I'll start pulling it into the v4_integration branch of plotly.py

nicolaskruchten commented on plotly_express/_core.py:
```diff
@@ -74,9 +62,17 @@ def get_trendline_results(fig):
+    ],
 )
-TraceSpec = namedtuple("TraceSpec", ["constructor", "attrs", "trace_patch"])
+TraceSpec = namedtuple("TraceSpec", ["constructor", "attrs", "trace_patch", "marginal"])
```
This is a good idea, and maybe should be generalized later to model the "role" of the trace in the figure... Is it the main event, or a marginal, or a trendline, or some other future thing like the raw scatter that someone might want to add to a density heatmap?

nicolaskruchten commented on plotly_express/_core.py:
```diff
@@ -302,126 +301,132 @@ def configure_axes(args, constructor, fig, axes, orders):
     range_key = "range_" + letter
     if log_key in args and args[log_key]:
-        layout[axis]["type"] = "log"
+        axis["type"] = "log"
```
if `axis` is a `go` then it would be nicer to have `axis.type` here right?

nicolaskruchten replied:
(not necessary, just noting)

nicolaskruchten commented on plotly_express/_core.py:
```diff
@@ -38,25 +45,6 @@ def set_mapbox_access_token(token):
     MAPBOX_TOKEN = token
 
 
-class ExpressFigure(go.Figure):
```
should grep the docstrings for `ExpressFigure` and change that out :)

nicolaskruchten commented:
This looks great! Having marginals vs facets work this way on a per-direction basis is really 👍 !!

nicolaskruchten commented:
Let's pull this into v4

jonmmease commented:
Thanks @nicolaskruchten, superseded by https://github.com/plotly/plotly.py/pull/1613
